### PR TITLE
v4.2.0: Dev to Trunk

### LINKED ISSUE #869: Admin/Instructor can't view checkout as self if they're enrolled in the course
### Reproduction Steps

+ Login to WP as an admin (or instructor)
+ Enroll yourself in a course
+ Attempt to review an access plan for the course on the frontend

### Expected Behavior

+ User should be able to use "View as" settings found on courses/memberships to view the checkout as a visitor / enrolled student

### Actual Behavior

+ User is displayed the enrolled notice via #867 (good)
+ user cannot preview the checkout via "View as" settings (https://lifterlms.com/docs/lifterlms-preview-management/)

### LINKED ISSUE #1039: Orphan quizzes (deletion) edge case
### Reproduction Steps

Reported by a user.

I don't have a way to reproduce this programmatically, but the idea is that one can have quizzes which are linked to a lesson but this lesson has not quizzes attached or has a different quiz attached.
The issue can be reproduced with the affected site's db: in LifterLMS -> Reporting -> Quizzes you could see more than one quiz "attached" to the same lesson but none of them displayed the Delete button

### Expected Behavior

I would expect that a lesson had only one quiz and that except for that one the others would be deletable.

### Actual Behavior

More than one quiz looked attached to a lesson, non of them displayed the Delete button.

### Error Messages / Logs

n/a

### System Information

LifterLMS 3.27.11

### Browser, Device, and Operating System Information

n/a

### Related User Information
HS-112131

### Technical Details

I've found that all the offending quizzes have the `lesson_id` meta still set to say `5`, but the lesson with id `5` had, of course, the `quiz_id` meta equal to only one of them.

My idea is that we should allow for a more in depth check on whether a quiz is an orphan or not, which means that not only its `lesson_id` must be unset, but that if it's set then its parent lesson's quiz must be equal to the this quiz id.
For this purpose I would add a `$deep` param to:
https://github.com/gocodebox/lifterlms/blob/3.37.11/includes/models/model.llms.quiz.php#L147
`false` by default (so to have the smaller possible impact), that allows for this `deep` check, and use it here:
https://github.com/gocodebox/lifterlms/blob/3.37.11/includes/admin/reporting/tables/llms.table.quizzes.php#L98
and wherever else is needed.

### LINKED ISSUE #1116: Course outline not unmarking lessons complete when quiz attempt removed
### Reproduction Steps

+ Add the course outline to a course
+ Add a quiz to the course
+ Take the quiz
+ Go into Reporting and remove quiz attempt
+ Clear student reporting cache
+ Go back into the course and look at the syllabus

### Expected Behavior

+ The lesson for the quiz should be marked as incomplete not crossed out

### Actual Behavior

+ it's still crossed out.

HS-118408

### LINKED ISSUE #1142: PHP Warning when deleting a quiz attempt
### Reproduction Steps

+ From LifterLMS -> Reporting -> Quizzes visit the Attempts screen (so you need at least a quiz with an attempt)
+ Click on a quiz attempt
+ Delete the attempt

### Expected Behavior

+ Attempt deleted with no warnings

### Actual Behavior

+ Attempt deleted but with warnings

### Error Messages / Logs

```
[12-May-2020 12:14:29 UTC] PHP Notice:  Undefined index:  in /srv/www/wordpress-one/public_html/wp-content/plugins/lifterlms/includes/models/model.llms.quiz.attempt.php on line 614

```

### System and Environment Information

```
Php Version: 7.2.29-1+ubuntu18.04.1+deb.sury.org+1


LifterLMS 3.38.1
```

This issue has be recreated:
+ [x] Locally
+ [ ] On a staging site
+ [ ] On a production website
+ [ ] With only LifterLMS and a default theme

### Browser, Device, and Operating System Information

n/a

### LINKED ISSUE #1156: LLMS_Abstract_User_Data::get_id() should always return an int
https://github.com/gocodebox/lifterlms/blob/dcc6bdf184661cf3e9cf5d36803cbff72f51f524/includes/abstracts/llms.abstract.user.data.php#L180-L189

Return value should be passed through `absint()`

### LINKED ISSUE #1227: Firefox says that the llms-tracking cookie will be rejected soon because of missing/incorrect attributes
### Reproduction Steps

+ With the FireFox Browser
+ Login as whatever user
+ Visit the front
+ Open the dev tools and look at the console

### Expected Behavior

+ I expect to see no llms cookie relates warnings

### Actual Behavior

+ I see a warning related to the llms-tracking cooke that will be rejected soon 

### Error Messages / Logs

+ Include any relevant error messages or log files
```
llms.js:1443 Cookie "llms-tracking" will be soon rejected because it has the "sameSite" attribute set to "none" or an invalid value, without the "secure" attribute. To know more about the "sameSite" attribute, read https://developer.mozilla.org/docs/Web/HTTP/Headers/Set-Cookie/SameSite

```

### System and Environment Information

LifterLMS 4.0.0

This issue has be recreated:
+ [x] Locally
+ [ ] On a staging site
+ [ ] On a production website
+ [x] With only LifterLMS and a default theme

### Browser, Device, and Operating System Information

+ Browser: Firefox 78.0.1

### LINKED ISSUE #1229: When setting a tooltip the icon doesn't load
### Reproduction Steps

+ Set a tooltip with an icon here: https://polished-wildflower.myliftersite.com/wp-admin/admin.php?page=llms-settings&tab=integrations&section=gravityview
+ Contact Zack Katz with any clarifying questions: https://github.com/zackkatz

### Expected Behavior

+ Icon loads with tooltip

### Actual Behavior

+ when adding a setting field with a tooltip ($field['desc_tooltip']), the icon doesn't load

### Error Messages / Logs

+ Error: https://www.evernote.com/l/AnjyKvljEQBLvLiLyc4fZUJyUAvUxpJmFQA
```
<!-- Paste error logs / backtraces below this line -->

```

### System and Environment Information

<details>
<summary>System Report</summary>

```
Wordpress
-------------------------------------------

Home Url: https://polished-wildflower.myliftersite.com
Site Url: https://polished-wildflower.myliftersite.com
Login Url: https://polished-wildflower.myliftersite.com/wp-login.php
Version: 5.4.2
Debug Mode: No
Debug Log: No
Debug Display: Yes
Locale: en_US
Multisite: No
Page For Posts: Blog (#1892) [https://polished-wildflower.myliftersite.com/blog/]
Page On Front: Welcome to Your LifterLMS 30 Day Demo! (#1772) [https://polished-wildflower.myliftersite.com/]
Permalink Structure: /%postname%/
Show On Front: page
Wp Cron: Yes


Settings
-------------------------------------------

Version: 4.1.0
Db Version: 4.1.0
Course Catalog: All Courses (#2305) [https://polished-wildflower.myliftersite.com/all-courses/]
Membership Catalog: Membership Catalog (#2158) [https://polished-wildflower.myliftersite.com/memberships/]
Student Dashboard: My Courses (#1774) [https://polished-wildflower.myliftersite.com/my-courses/]
Checkout Page: Purchase (#2159) [https://polished-wildflower.myliftersite.com/purchase/]
Course Catalog Per Page: 20
Course Catalog Sorting: menu_order,ASC
Membership Catalog Per Page: 9
Membership Catalog Sorting: menu_order
Site Membership: Not Set
Courses Endpoint: my-courses
Edit Endpoint: edit-account
Lost Password Endpoint: lost-password
Vouchers Endpoint: redeem-voucher
Autogenerate Username: yes
Password Strength Meter: no
Minimum Password Strength: weak
Terms Required: no
Terms Page: Not Set
Checkout Names: required
Checkout Address: hidden
Checkout Phone: hidden
Checkout Email Confirmation: no
Open Registration: no
Registration Names: required
Registration Address: optional
Registration Phone: hidden
Registration Voucher: optional
Registration Email Confirmation: no
Account Names: required
Account Address: required
Account Phone: optional
Account Email Confirmation: yes
Confirmation Endpoint: confirm-payment
Force Ssl Checkout: no
Country: US
Currency: USD
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Decimals: 2
Trim Zero Decimals: no
Recurring Payments: yes
Email From Address: test@test.com
Email From Name: Sandbox
Email Footer Text: 
Email Header Image: 
Cert Bg Width: 800
Cert Bg Height: 616
Cert Legacy Compat: no


Gateways
-------------------------------------------

Stripe: Enabled
Stripe Test Mode: Enabled
Stripe Logging: yes
Stripe Order: 1
Manual: Enabled
Manual Logging: yes
Manual Order: 2
PayPal: Enabled
PayPal Test Mode: Enabled
PayPal Logging: yes
PayPal Order: 3
Authorize.Net: Disabled
Authorize.Net Test Mode: Enabled
Authorize.Net Logging: yes
Authorize.Net Order: 4


Server
-------------------------------------------

Mysql Version: 5.7.30
Php Curl: Yes
Php Default Timezone: UTC
Php Fsockopen: Yes
Php Max Input Vars: 5000
Php Max Upload Size: 512 MB
Php Memory Limit: 256M
Php Post Max Size: 1024M
Php Soap: Yes
Php Suhosin: No
Php Time Limt: 30
Php Version: 7.3.19
Software: Apache/2.4.43 (Unix) OpenSSL/1.1.1
Wp Memory Limit: 256M


Browser
-------------------------------------------

HTTP USER AGENT: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/83.0.4103.116 Safari/537.36


Theme
-------------------------------------------

Name: Kadence
Version: 0.7.11
Themeuri: https://www.kadencewp.com/product/kadence/
Authoruri: https://www.kadencewp.com/
Template: 
Child Theme: No
Llms Support: No


Plugins
-------------------------------------------

Gravity Forms: 2.4.18
GravityView: 2.9
GravityView - Advanced Filter Extension: 2.1
LifterLMS: 4.1.0
LifterLMS Advanced Quizzes: 1.0.10
LifterLMS Advanced Videos: 1.0.0-beta.10
LifterLMS Assignments: 1.1.9
LifterLMS Authorize.Net Payment Gateway: 1.0.5
LifterLMS ConvertKit: 2.1.0
LifterLMS Gravity Forms: 2.1.1
LifterLMS MailChimp Add-on: 3.1.2
LifterLMS PayPal Gateway: 1.3.0
LifterLMS PDFs: 1.0.0
LifterLMS Private Areas: 1.1.2
LifterLMS Social Learning: 1.2.1
LifterLMS Stripe Payment Gateway: 5.2.0


Integrations
-------------------------------------------

BbPress: No
BuddyPress: No
GravityView: Yes
LifterLMS ConvertKit: No
LifterLMS Gravity Forms: Yes
LifterLMS MailChimp: No
LifterLMS Private Areas: No
LifterLMS Social Learning: No
Videos: Vimeo: No
Videos: Wistia: Yes
Videos: YouTube: No


Template Overrides
-------------------------------------------

course/lesson-preview.php (ver: 3.19.2): /srv/users/polished-wildflower/apps/polished-wildflower/public/wp-content/themes/kadence/lifterlms/ (ver: 3.19.2)
course/syllabus.php (ver: 3.24.0): /srv/users/polished-wildflower/apps/polished-wildflower/public/wp-content/themes/kadence/lifterlms/ (ver: 3.24.0)



```

</details>

This issue has be recreated:
+ [ ] Locally
+ [ ] On a staging site
+ [ ] On a production website
+ [ ] With only LifterLMS and a default theme

### Browser, Device, and Operating System Information

+ Browser name and version
+ Operating System name and version
+ Device name and version (if applicable)

### LINKED ISSUE #1236: Preview management wp-admin menu icon uses old logo
### Reproduction Steps

+ Log in as an admin
+ Visit the frontend of course
+ Look at the "View as" menu on the top right of the WP Admin bar

### Expected Behavior

+ See the "new" logo as the icon

### Actual Behavior

+ The old icon is visible

### PR DESCRIPTION
**Pull request by thomasplevy**
+ Fixes #869 via #874 
+ Fixes #1039 via #1240 
+ Fixes #1116 and Fixes #1142 via #1237 
+ Fixes #1156 via #1243 
+ Fixes #1227 via #1234 
+ Fixes #1229 via #1230
+ Fixes #1236 via #1245 
+ PR #1222
+ PR #1233 (thanks @nicolas-jaussaud)
+ PR #1242 
+ PR #1244 

Pending release asset: [lifterlms-4.2.0.zip](https://github.com/gocodebox/lifterlms/files/4950218/lifterlms-4.2.0.zip)

### CODE REVIEW

**Comment by thomasplevy** [comment]
I can't find a *good* way to reduce method length for the 2 issues reported by CodeClimate.

**Comment by thomasplevy** [comment]
Pending release asset

[lifterlms-4.2.0.zip](https://github.com/gocodebox/lifterlms/files/4950214/lifterlms-4.2.0.zip)

**Comment by thomasplevy** [comment]
@eri-trabiccolo All CI tests are passing, preflight manual tests look good, thumbs up from you and I'll ship it tomorrow

File: includes/admin/post-types/meta-boxes/class.llms.meta.box.order.enrollment.php [hunk 1]
```diff
@@ -15,7 +15,7 @@
  *
  * @since 3.0.0
  * @since 3.33.0 Added the logic to handle the Enrollment 'deleted' status on save.
- * @since [version] In ` save_delete_enrollment()` removed order cancellation instruction, moved elsewhere as reaction to the enrollment deletion.
+ * @since 4.2.0 In ` save_delete_enrollment()` removed order cancellation instruction, moved elsewhere as reaction to the enrollment deletion.
  *                @see `LLMS_Controller_Orders->on_deleted_enrollment()` in `includes\controllers\class.llms.controller.orders.php`.
  *                Also, add order note about the enrollment deletion only if it actually occurred.
```
Comment (eri-trabiccolo) [inline includes/admin/post-types/meta-boxes/class.llms.meta.box.order.enrollment.php]:
```suggestion
 *              @see `LLMS_Controller_Orders->on_deleted_enrollment()` in `includes\controllers\class.llms.controller.orders.php`.
 *              Also, add order note about the enrollment deletion only if it actually occurred.
```

File: includes/admin/post-types/meta-boxes/class.llms.meta.box.order.enrollment.php [hunk 2]
```diff
@@ -136,7 +136,7 @@ public function save( $post_id ) {
 	 * Delete enrollment data based on posted values.
 	 *
 	 * @since 3.33.0
-	 * @since [version] Removed order cancellation instruction, moved elsewhere as reaction to the enrollment deletion.
+	 * @since 4.2.0 Removed order cancellation instruction, moved elsewhere as reaction to the enrollment deletion.
 	 *                @see `LLMS_Controller_Orders->on_deleted_enrollment()` in `includes\controllers\class.llms.controller.orders.php`.
 	 *                Also, add order note about the enrollment deletion only if it actually occurred.
```
Comment (eri-trabiccolo) [inline includes/admin/post-types/meta-boxes/class.llms.meta.box.order.enrollment.php]:
```suggestion
	 *              @see `LLMS_Controller_Orders->on_deleted_enrollment()` in `includes\controllers\class.llms.controller.orders.php`.
	 *              Also, add order note about the enrollment deletion only if it actually occurred.
```

File: includes/class.llms.view.manager.php [hunk 3]
```diff
@@ -17,7 +17,7 @@
  *
  * @since 3.7.0
  * @since 3.35.0 Sanitize `$_GET` data.
- * @since [version] Disable the view management when creating a pending order.
+ * @since 4.2.0 Disable the view management when creating a pending order.
  *               Added `modify_display_free_enroll_form` method.
  *               `get_url` method modified so to take into account already present
  *               query args in the request URI.
```
Comment (eri-trabiccolo) [inline includes/class.llms.view.manager.php]:
```suggestion
 *              Added `modify_display_free_enroll_form` method.
 *              `get_url` method modified so to take into account already present
 *              query args in the request URI.
```

File: includes/controllers/class.llms.controller.orders.php [hunk 4]
```diff
@@ -18,7 +18,7 @@
  * @since 3.34.4 Added filter `llms_order_can_be_confirmed`.
  * @since 3.34.5 Fixed logic error in `llms_order_can_be_confirmed` conditional.
  * @since 3.36.1 In `recurring_charge()`, made sure to process only proper LLMS_Orders of existing users.
- * @since [version] Added logic to set the order status to 'cancelled' when an enrollment linked to an order is deleted.
+ * @since 4.2.0 Added logic to set the order status to 'cancelled' when an enrollment linked to an order is deleted.
  *               Also `llms_unenroll_on_error_order` fiter hook added.
```
Comment (eri-trabiccolo) [inline includes/controllers/class.llms.controller.orders.php]:
```suggestion
 *              Also `llms_unenroll_on_error_order` fiter hook added.
```

File: includes/controllers/class.llms.controller.orders.php [hunk 5]
```diff
@@ -318,7 +318,7 @@ public function create_pending_order() {
 	 *
 	 * @since 3.0.0
 	 * @since 3.10.0 Unknown.
-	 * @since [version] Added `llms_unenroll_on_error_order` filter hook.
+	 * @since 4.2.0 Added `llms_unenroll_on_error_order` filter hook.
 	 *
 	 * @param  LLMS_Order $order Instance of an LLMS_Order
```
Comment (eri-trabiccolo) [inline includes/controllers/class.llms.controller.orders.php]:
```suggestion
	 * @param LLMS_Order $order Instance of an LLMS_Order.
```

File: includes/models/model.llms.student.php [hunk 6]
```diff
@@ -1777,7 +1777,7 @@ public function delete_enrollment( $product_id, $trigger = 'any' ) {
 	 * Inserts / updates necessary user postmeta data.
 	 *
 	 * @since 3.17.0
-	 * @since [version] Use filterable functions to determine if the object is completable.
+	 * @since 4.2.0 Use filterable functions to determine if the object is completable.
 	 *                Added filter to allow customization of object parent data.
```
Comment (eri-trabiccolo) [inline includes/models/model.llms.student.php]:
```suggestion
	 *              Added filter to allow customization of object parent data.
```

**Review by eri-trabiccolo — CHANGES_REQUESTED** [review]
I just noticed some minor doc cs issues, I don't know if you want to fix them now.
For the rest **thumbs up**, ship it.

File: includes/class.llms.view.manager.php (hunk 3 above)
Comment (eri-trabiccolo) [reply includes/class.llms.view.manager.php]: not sure why this and the other two above are labeled as "outdated"...

**Pull request merged**